src/com/google/javascript/jscomp/AdvancedCoverageInstrumentationCallback.java

```diff
@@ -0,0 +1,205 @@
+        compiler.reportChangeToEnclosingScope(node);
+        instrumentCode(traversal, instrumentationData.get(fileName));
+      }
+    }
+
+    if (node.isFunction()) {
+
+      functionName = NodeUtil.getBestLValueName(NodeUtil.getBestLValue(node));
+
+      if (!instrumentationData.containsKey(fileName)) {
+        instrumentationData.put(fileName, new AdvancedFileInstrumentationData(fileName));
+      }
+      addBlockNode(instrumentationData.get(fileName), Arrays.asList(node.getLastChild()),
+          functionName);
+
+    }
+  }
+
+  /**
+   * Iterate over all collected block nodes within a Script node and add a new child to the front of
+   * each block node which is the instrumentation Node
+   *
+   * @param traversal Tne node traversal context which maintains information such as fileName being
+   *                  traversed
+   * @param data      Data structure that maintains an organized list of block nodes that need to be
+   *                  instrumented
+   */
+  private void instrumentCode(NodeTraversal traversal, AdvancedFileInstrumentationData data) {
+    Map<String, List<Node>> functionScopedNodes = data.getFunctionScopeNodes();
+    for (String key : functionScopedNodes.keySet()) {
+      for (Node block : functionScopedNodes.get(key)) {
+        block.addChildToFront(
+            newInstrumentationNode(traversal, block, key));
+        compiler.reportChangeToEnclosingScope(block);
+      }
+    }
+  }
+
+  /**
+   * Create a function call to the Instrument Code function with properly encoded parameters.
+   *
+   * @param traversal The context of the current traversal.
+   * @param node      The block node to be instrumented.
+   * @param fnName    The function name that the node exists within.
+   * @return The newly constructed function call node.
+   */
+  private Node newInstrumentationNode(NodeTraversal traversal, Node node, String fnName) {
+
+    String type = "Type.FUNCTION";
+    String combinedParam = traversal.getSourceName() + " " + fnName + " " + type;
+
+    BigInteger uniqueIdentifier = parameterMapping.getUniqueIdentifier(combinedParam);
+    BigInteger maxInteger = new BigInteger(Integer.toString(Integer.MAX_VALUE));
+
+    if (uniqueIdentifier.compareTo(maxInteger) > 0) {
+      throw new ArithmeticException(
+          "Unique Identifier exceeds value of Integer.MAX_VALUE, could not encode with Base 64 VLQ");
+    }
+
+    StringBuilder sb = new StringBuilder();
+
+    try {
+      Base64VLQ.encode(sb, uniqueIdentifier.intValue());
+    } catch (IOException e) {
+      // If encoding in Base64 VLQ fails, we will use the original identifier as it will still
+      // maintain obfuscation and partial optimization
+      sb.append(uniqueIdentifier.intValue());
+    }
+
+    parameterMapping.addParamMapping(sb.toString(), combinedParam);
+
+    Node inner_prop = IR
+        .getprop(IR.name("module$exports$instrument$code"), IR.string("instrumentCodeInstance"));
```

Comment:
Could these also be static constants? I think it would be useful to comment on where this function name comes from, too.